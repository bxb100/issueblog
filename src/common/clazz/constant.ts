export class Constant {
    static readonly LINKS = 'Links'
    static readonly TOP = 'Top'
    static readonly RECENT = 'Recent'
    static readonly TODO = 'Todo'
    static readonly EACH_LABEL = 'Label'

    readonly header: string

    constructor(header: string) {
        this.header = header
    }

    convertBlogContent(map: Map<string, string>): string {
        let content: string = this.header
        content += map.get(Constant.LINKS) || ''
        content += map.get(Constant.TOP) || ''
        content += map.get(Constant.RECENT) || ''
        content += map.get(Constant.EACH_LABEL) || ''
        content += map.get(Constant.TODO) || ''
        return content
    }
}
