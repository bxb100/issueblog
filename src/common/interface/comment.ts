import {IUser} from './user'
import {Reactions} from './reactions'

export interface IComment {
    id: number
    user: IUser | null
    body?: string
    reactions?: Reactions
}
