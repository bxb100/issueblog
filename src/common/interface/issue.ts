import {IAssignee} from './assignee'
import {ILabel} from './label'
import {IMilestone} from './milestone'
import {IUser} from './user'
import {IsoDateString} from '../types/iso-date-string'

export interface IIssue {
    title: string
    number: number
    created_at: IsoDateString
    updated_at: IsoDateString
    labels: (string | ILabel)[]
    pull_request?: Object | null
    state: string
    locked: boolean
    milestone?: IMilestone | null
    user: IUser | null
    assignees?: IAssignee[] | null
    comments: number
    html_url: string
    body?: string | null
}
