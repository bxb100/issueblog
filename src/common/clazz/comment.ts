import {GithubKit} from './github-kit'
import {IComment} from '../interface/comment'
import {IUser} from '../interface/user'
import {Reactions} from '../interface/reactions'

export class Comment implements IComment {
    readonly id: number
    readonly user: IUser | null
    readonly reactions: Reactions | undefined
    readonly body: string | undefined

    constructor(comment: Readonly<IComment>) {
        this.id = comment.id
        this.user = comment.user
        this.body = comment.body
        this.reactions = comment.reactions
    }

    static cast(comments: Readonly<IComment[]>): Comment[] {
        return comments.map((c: IComment): Comment => new Comment(c))
    }

    async isHeartBySelf(util: GithubKit): Promise<boolean> {
        if (this._existHeartReaction()) {
            return util.isHeartBySelf(this)
        }
        return Promise.resolve(false)
    }

    private _existHeartReaction(): boolean {
        return (this.reactions?.heart || 0) >= 1
    }
}
