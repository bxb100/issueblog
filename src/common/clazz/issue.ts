import {GithubKit} from './github-kit'
import {IAssignee} from '../interface/assignee'
import {IIssue} from '../interface/issue'
import {ILabel} from '../interface/label'
import {IMilestone} from '../interface/milestone'
import {IUser} from '../interface/user'
import {IsoDateString} from '../types/iso-date-string'

export class Issue implements IIssue {
    readonly assignees: IAssignee[] | null | undefined
    readonly comments: number
    readonly created_at: IsoDateString
    readonly created_at_sub: string
    readonly labels: (string | ILabel)[]
    readonly locked: boolean
    readonly milestone: IMilestone | null | undefined
    readonly number: number
    readonly pull_request: Object | null | undefined
    readonly state: string
    readonly title: string
    readonly updated_at: IsoDateString
    readonly user: IUser | null
    readonly html_url: string
    readonly body: string | null | undefined

    constructor(data: IIssue) {
        this.assignees = data.assignees
        this.comments = data.comments
        this.created_at = data.created_at
        this.labels = data.labels
        this.locked = data.locked
        this.milestone = data.milestone
        this.number = data.number
        this.pull_request = data.pull_request
        this.state = data.state
        this.title = data.title
        this.updated_at = data.updated_at
        this.user = data.user
        this.html_url = data.html_url
        this.body = data.body

        // easy way get yyyy-MM-dd
        this.created_at_sub = this.created_at.substring(0, 10)
    }

    static cast(data: IIssue[]): Issue[] {
        return data.map(d => new Issue(d))
    }

    static getLabelValue(label: string | ILabel): string | undefined {
        if (typeof label === 'object') {
            return label.name
        }
        return label
    }

    getLabelName(kit: GithubKit): string[] {
        return this.labels
            .map(l => Issue.getLabelValue(l) || kit.config.unlabeled_title)
            .filter(Boolean)
    }

    containLabel(label: string): boolean {
        label = label.toLowerCase().trim()
        return this.labels.some(
            l => Issue.getLabelValue(l)?.toLowerCase() === label
        )
    }

    bodyToLines(): string[] {
        return this.body?.split('\n').map(line => line.trim()) || []
    }

    mdIssueInfo(): string {
        return `- [${this.title}](${this.html_url})---${this.created_at_sub}\n`
    }
}
