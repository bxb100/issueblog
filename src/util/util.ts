import {IComment} from '../common/interface/comment'
import {IIssue} from '../common/interface/issue'
import {IsoDateString} from '../common/types/iso-date-string'
import {Issue} from '../common/clazz/issue'
import {MetadataInfo} from '../common/types/metadata'
import {WrapperToString} from '../common/types/wrapper-to-string'

export function isOwnBy(obj: IIssue | IComment, username: string): boolean {
    return obj.user?.login === username
}

export function wrapDetails<T>(
    shows: T[],
    hides: T[],
    wrapper: WrapperToString<T>
): string {
    let result = ''

    if (shows.length !== 0) {
        result += shows.map(wrapper).join('')
    }

    if (hides.length !== 0) {
        result += '<details><summary>MORE</summary>'
        result += '\n\n'

        result += hides.map(wrapper).join('')

        result += '</details>'
        result += '\n'
    }
    result += '\n'
    return result
}

export function backupFileName(issue: Issue): string {
    return `${issue.number}-${issue.title.replace(/\t|\s/g, '_')}.md`
}

export function compareUpdateTime(a: MetadataInfo, b: IsoDateString): number {
    return new Date(a.updatedAt).getTime() - Date.parse(b)
}

export function unifyReferToNumber(issue: Issue, comments: IComment[]): string {
    let n = 1
    const referenceLinks: string[] = []

    function replaceReferenceLink(content: Issue | IComment): void {
        if (!content.body) return
        // remove all the reference links
        for (const ref of content.body.match(/\[\^\d](?!:)/gm) || []) {
            const number = ref.slice(2, -1)
            const rg = RegExp(`\\[\\^${number}]:\\s*.+`, 'gm')
            // eslint-disable-next-line github/array-foreach
            content.body.match(rg)?.forEach(link => {
                content.body = content.body?.replace(link, '')
                referenceLinks.push(link.replace(/(^\[\^\d+]:\s*)/, ''))
            })
            content.body = content.body.replace(ref, `[^${n++}]`)
        }
    }

    replaceReferenceLink(issue)

    for (const comment of comments) {
        replaceReferenceLink(comment)
    }

    return referenceLinks.reduce((p, c, index) => {
        return `${p}\n[^${index + 1}]: ${c}`
    }, '')
}
