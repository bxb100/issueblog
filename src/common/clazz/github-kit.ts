import * as core from '@actions/core'
import {context, getOctokit} from '@actions/github'
import {Comment} from './comment'
import {Config} from '../../util/config'
import {Constant} from './constant'
import {GitHub} from '@actions/github/lib/utils'
import {IComment} from '../interface/comment'
import {IIssue} from '../interface/issue'
import {IRelease} from '../interface/release'
import {Issue} from './issue'
import {Reaction} from '../interface/reaction'
import {ReactionContent} from '../enum/reaction-content'
import {Release} from './release'
import {add_md_friends} from '../../functions/links-process'
import {add_md_label} from '../../functions/label-process'
import {add_md_recent} from '../../functions/recent-process'
import {add_md_todo} from '../../functions/todo-process'
import {add_md_top} from '../../functions/top-process'
import {backup} from '../../functions/backup'
import fs from 'fs'
import {rss} from '../../functions/rss'

export class GithubKit {
    readonly client: InstanceType<typeof GitHub>
    readonly owner: string
    readonly repo: string
    readonly config: Config
    readonly sectionMap = new Map<string, string>()

    constructor(config: Config) {
        this.config = config
        this.client = getOctokit(config.github_token)
        this.owner = context.repo.owner
        this.repo = context.repo.repo
    }

    async isHeartBySelf(comment: IComment): Promise<boolean> {
        const reactions = await this.getIssueCommentReactions(
            comment,
            ReactionContent.HEART
        )
        return !!reactions.find(r => r.user?.login === this.owner)
    }

    async getIssueCommentReactions(
        comment: IComment,
        content?: ReactionContent
    ): Promise<Reaction[]> {
        const reactions = await this.client.rest.reactions.listForIssueComment({
            owner: this.owner,
            repo: this.repo,
            comment_id: comment.id,
            content
        })
        core.debug(`reactions:\n\n${JSON.stringify(reactions)}\n\n`)
        return reactions.data
    }

    async getIssueComments(issue: IIssue): Promise<Comment[]> {
        const comments = await this.client.rest.issues.listComments({
            owner: this.owner,
            repo: this.repo,
            issue_number: issue.number,
            per_page: 100
        })
        core.debug(`comments:\n\n${JSON.stringify(comments)}\n\n`)
        return Comment.cast(comments.data)
    }

    async getIssues(page: number): Promise<Issue[]> {
        const issueResult = await this.client.rest.issues.listForRepo({
            owner: this.owner,
            repo: this.repo,
            state: 'open',
            creator: this.owner,
            per_page: 100,
            direction: 'desc',
            page
        })
        core.debug(`issueResult:\n\n${JSON.stringify(issueResult.data)}\n\n`)
        return Issue.cast(issueResult.data)
    }

    async getRelease(page: number): Promise<IRelease[]> {
        const releases = await this.client.rest.repos.listReleases({
            owner: this.owner,
            repo: this.repo,
            per_page: 100,
            page
        })
        core.debug(`releases:\n\n${JSON.stringify(releases)}\n\n`)
        return releases.data
    }

    async getAllReleases(): Promise<Release[]> {
        let page = 1
        let releases: IRelease[] = []
        // eslint-disable-next-line no-constant-condition
        while (true) {
            const result = await this.getRelease(page)
            if (result.length === 0) {
                break
            }
            releases = releases.concat(result)
            page++
        }
        return Release.cast(releases)
    }

    async renderMarkdown(markdown: string): Promise<string> {
        // noinspection ES6RedundantAwait,TypeScriptValidateJSTypes
        const rendered = await this.client.rest.markdown.render({
            text: markdown,
            mode: 'markdown'
        })
        core.debug(`rendered:\n\n${JSON.stringify(rendered)}\n\n`)
        return rendered.data
    }

    async getAllIssues(): Promise<Issue[]> {
        let page = 1
        let issues: Issue[] = []
        // eslint-disable-next-line no-constant-condition
        while (true) {
            const result = await this.getIssues(page)
            if (result.length === 0) {
                break
            }
            issues = issues.concat(result)
            page++
        }
        return issues
    }

    async process(): Promise<void[]> {
        // using subscriber-publisher rewrite this is better?
        const issues = await this.getAllIssues()

        const mainProcess: Promise<void> = Promise.all([
            add_md_friends(this, issues),
            add_md_top(this, issues),
            add_md_recent(this, issues),
            add_md_todo(this, issues),
            add_md_label(this, issues)
        ]).then(
            () => {
                const constant = new Constant(this.config.md_header)
                fs.writeFileSync(
                    'README.md',
                    constant.convertBlogContent(this.sectionMap)
                )
            },
            err => {
                throw err
            }
        )
        return Promise.all([
            mainProcess,
            rss(this, issues),
            backup(this, issues)
        ])
    }
}
