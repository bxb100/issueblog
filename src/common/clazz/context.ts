import {GithubKit} from './github-kit'
import {Constant} from './constant'
import {Issue} from './issue'
import {Config} from '../../util/config'

export class Context {
    readonly kit: GithubKit
    readonly issues: Issue[] = []
    readonly sectionMap = new Map<string, string>()
    readonly essayIssues: Issue[] = []
    readonly config: Config

    constructor(issues: Issue[], kit: GithubKit, config: Config) {
        this.kit = kit
        this.issues = issues
        this.essayIssues = Context.getEssayIssues(issues)
        this.config = config
    }

    getIssues(label: string): Issue[] {
        if (label) {
            return this.issues.filter(issue => issue.hasLabel(label))
        }
        return this.issues
    }

    private static getEssayIssues(issues: Issue[]): Issue[] {
        return issues.filter(
            issue =>
                !(
                    issue.hasLabel(Constant.FIXED_LINKS) ||
                    issue.hasLabel(Constant.FIXED_TODO)
                )
        )
    }
}
