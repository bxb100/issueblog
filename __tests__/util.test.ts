// @ts-ignore
import fs from 'fs'
import {Issue} from '../src/common/clazz/issue'
import {Comment} from '../src/common/clazz/comment'
import {unifyReferToNumber} from '../src/util/util'

test('test unifyReferToNumber', async () => {
    const issue: Issue = new Issue(
        JSON.parse(fs.readFileSync('./__tests__/mock/refer-issue.json', 'utf8'))
    )

    const comments = Comment.cast(
        JSON.parse(
            fs.readFileSync(
                './__tests__/mock/refer-issue-comments.json',
                'utf8'
            )
        )
    )

    expect(unifyReferToNumber(issue, comments)).toMatchSnapshot()
    expect(issue.body).toMatchSnapshot()
    for (let comment of comments) {
        expect(comment.body).toMatchSnapshot()
    }
})

