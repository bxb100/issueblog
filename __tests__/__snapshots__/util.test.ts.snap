// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test unifyReferToNumber 1`] = `
"
[^1]: https://eisel.me/lld
[^2]: https://www.reddit.com/r/rust/comments/11h28k3/faster_apple_builds_with_the_lld_linker/
[^3]: https://github.com/BurtonQin/lockbud/issues/44
[^4]: https://github.com/actix/actix-web/issues/1147#issuecomment-1509937750"
`;

exports[`test unifyReferToNumber 2`] = `
"~~## ZLD 的配置~~

~~Rust 编译大部分耗时在 [linker 阶段](https://en.wikipedia.org/wiki/Linker_(computing)), 所以文中给与 ZLD 配置, 但是注意后面的配置路径有点问题, homebrew Apple silicon 默认的安装位置在\`/opt/homebrew/bin/\`~~

\`\`\`diff
# .cargo/config.toml
# On Windows
# \`\`\`
# cargo install -f cargo-binutils
# rustup component add llvm-tools-preview
# \`\`\`
[target.x86_64-pc-windows-msvc]
rustflags = ["-C", "link-arg=-fuse-ld=lld"]
[target.x86_64-pc-windows-gnu]
rustflags = ["-C", "link-arg=-fuse-ld=lld"]
# On Linux:
# - Ubuntu, \`sudo apt-get install lld clang\`
# - Arch, \`sudo pacman -S lld clang\`
[target.x86_64-unknown-linux-gnu]
rustflags = ["-C", "linker=clang", "-C", "link-arg=-fuse-ld=lld"]
# On MacOS, \`brew install michaeleisel/zld/zld\`
[target.x86_64-apple-darwin]
rustflags = ["-C", "link-arg=-fuse-ld=/usr/local/bin/zld"]
[target.aarch64-apple-darwin]
- rustflags = ["-C", "link-arg=-fuse-ld=/usr/local/bin/zld"]
+ rustflags = ["-C", "link-arg=-fuse-ld=/opt/homebrew/bin/zld"]
\`\`\`
提到了新出来的 [mold](https://github.com/rui314/mold) 会更加好一点, 有机会再试试

---

update:
* zld 不再维护的了, 使用 lld 替代[^1], 但是和 reddit 上有用户评论一样, 速度并没有提示很高[^2]
* mold 的 MacOS 版本 sold 是商业版需要购买 licence , 放弃

总结: 不要花时间在 linker 上




"
`;

exports[`test unifyReferToNumber 3`] = `
"## CI 配置

* lockbud, 注意复制 tool-chain[^3]
* [GitHub Action yaml](https://gist.github.com/LukeMathWalker/5ae1107432ce283310c3e601fac915f3)

"
`;

exports[`test unifyReferToNumber 4`] = `
"## 技巧

* 使用 \`+night\` 不需要在项目中设置 \`tool-chain\`, 当然要注意下编译使用的版本需要和项目的一致, 比如 https://github.com/BurtonQin/lockbud
\`\`\`shell
# Use the nightly toolchain just for this command invocation
cargo +nightly expand
\`\`\`

* [actix-web](https://github.com/actix/actix-web) 支持共享 app 配置了[^4]
\`\`\`rust
fn create_app() -> App<
    impl ServiceFactory<
        ServiceRequest,
        Config = (),
        Response = ServiceResponse<impl MessageBody>,
        Error = Error,
        InitError = (),
    >,
> {
    App::new()
        .route("/", web::get().to(greet))
        .route("/health_check", web::get().to(health_check))
        .route("/{name}", web::get().to(greet))
}
\`\`\`


"
`;
